generic_fractional_dao/ligo/src/fractional_dao.mligo

```diff
@@ -0,0 +1,215 @@
+#if !FRACTIONAL_DAO
+#define FRACTIONAL_DAO
+
+#include "fa2_single_token.mligo"
+
+type permit = 
+[@layout:comb]
+{
+  key : key; (* user's key *)
+  signature : signature; (*signature of packed lambda + permit context *)
+}
+
+type proposal_info = {
+  vote_amount : nat;
+  voters : address set;
+  timestamp : timestamp;
+}
+
+type dao_lambda = unit -> operation list
+
+type vote =
+[@layout:comb]
+{
+  lambda : dao_lambda;
+  permit : permit option;
+}
+
+type set_voting_threshold_param = 
+{
+  old_threshold: nat;
+  new_threshold: nat;
+}
+
+type set_voting_period_param = 
+{
+  old_period: nat;
+  new_period: nat;
+}
+
+type pending_proposals = (bytes, proposal_info) big_map
+
+type dao_storage = {
+  ownership_token : single_token_storage;
+  voting_threshold : nat;
+  voting_period : nat;
+  vote_count : nat;
+  pending_proposals: pending_proposals;
+  metadata : contract_metadata;
+}
+
+type dao_entrypoints =
+  | Fa2 of fa2_entry_points
+  | Vote of vote
+  (** self-governance entry point *)
+  | Set_voting_threshold of set_voting_threshold_param
+  (** self-governance entry point *)
+  | Set_voting_period of set_voting_period_param
+  (** self-governance entry point *)
+  | Flush_expired of dao_lambda
+
+type return = (operation list) * dao_storage
+
+[@inline]
+let assert_self_call () =
+  if Tezos.sender = Tezos.self_address
+  then unit
+  else failwith "UNVOTED_CALL"
+
+let set_voting_threshold (t, s : set_voting_threshold_param * dao_storage)
+    : dao_storage =
+  if t.old_threshold <> s.voting_threshold
+  then (failwith "INVALID_OLD_THRESHOLD" : dao_storage)
+  else if t.new_threshold > s.ownership_token.total_supply
+  then (failwith "THRESHOLD_EXCEEDS_TOTAL_SUPPLY" : dao_storage)
+  else { s with voting_threshold = t.new_threshold; }
+
+let set_voting_period (p, s : set_voting_period_param * dao_storage)
+    : dao_storage =
+  if p.old_period <> s.voting_period
+  then (failwith "INVALID_OLD_PERIOD" : dao_storage)
+  else if p.new_period < 300n
+  then (failwith "PERIOD_TOO_SHORT" : dao_storage)
+  else { s with voting_period = p.new_period; }
+
+let is_expired (proposal, voting_period : proposal_info * nat) : bool =
+  if Tezos.now - proposal.timestamp > int(voting_period)
+  then true
+  else false
+
+let flush_expired (lambda, s : dao_lambda * dao_storage ) : dao_storage =
+  let key = Bytes.pack lambda in
+  match Big_map.find_opt key s.pending_proposals with
+  | None -> (failwith "PROPOSAL_DOES_NOT_EXIST" : dao_storage)
+  | Some proposal ->
+    if is_expired(proposal, s.voting_period)
+    then 
+      let new_pending = Big_map.remove key s.pending_proposals in
+      { s with pending_proposals = new_pending; }
+    else (failwith "NOT_EXPIRED" : dao_storage)
+
+
+let validate_permit (lambda, permit, vote_count 
+    : dao_lambda * permit * nat) : address =
+  let signed_data = Bytes.pack (
+    (Tezos.chain_id, Tezos.self_address),
+    (vote_count, lambda)
```

Comment:
This doesn't follow exactly the signing procedure in Tzip-17(https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-17/tzip-17.md#submission) where the parameter is packed and then hashed with `BLAKE2B` but I suppose that is unnecessary in this context since the un-obfuscated parameter is passed to the entrypoint at the same time anyway